sycl/include/sycl/vector.hpp

```diff
@@ -111,24 +97,807 @@ template <typename T> class GetOp {
+
+template <typename Lhs, typename Rhs, typename DataT>
+struct __SYCL_EBO NonTemplateBinaryOpAssignOpsMixin
+    : __SYCL_PROCESS_BINARY_OPASSIGN_OPS(__SYCL_MIXIN_FOR_BINARY_OPASSIGN,
+                                         __SYCL_COMMA) {};
+
+template <typename Self, typename VecT, typename DataT, int N>
+struct __SYCL_EBO SwizzleTemplateBinaryOpsMixin
+    : __SYCL_PROCESS_BINARY_OPS(__SYCL_MIXIN_FOR_TEMPLATE_BINARY,
+                                __SYCL_COMMA) {};
+
+template <typename Self, typename VecT, typename DataT, int N>
+struct __SYCL_EBO SwizzleTemplateBinaryOpAssignOpsMixin
+    : __SYCL_PROCESS_BINARY_OPASSIGN_OPS(
+          __SYCL_MIXIN_FOR_TEMPLATE_BINARY_OPASSIGN, __SYCL_COMMA) {};
+
+template <typename T, typename Impl, typename DataT>
+struct __SYCL_EBO UnaryOpsMixin
+    : __SYCL_PROCESS_UNARY_OPS(__SYCL_MIXIN_FOR_UNARY, __SYCL_COMMA) {};
+// clang-format on
+
+#undef __SYCL_MIXIN_FOR_UNARY
+#undef __SYCL_MIXIN_FOR_TEMPLATE_BINARY_OPASSIGN
+#undef __SYCL_MIXIN_FOR_BINARY_OPASSIGN
+#undef __SYCL_MIXIN_FOR_TEMPLATE_BINARY
+#undef __SYCL_MIXIN_FOR_BINARY
+
+#undef __SYCL_COMMA
+#undef __SYCL_PROCESS_BINARY_OPS
+#undef __SYCL_PROCESS_UNARY_OPS
+
+// Implement `<typename Impl>` parameters for the mixins above. These differ
+// between vec and swizzle.
+
+// Swizzle-specific part of the mixins' implementation.
+struct SwizzleImpl {
+private:
+  template <typename T> static constexpr int num_elements() {
+    if constexpr (is_vec_or_swizzle_v<T>)
+      return T::size();
+    else
+      return 1;
+  }
+
+public:
+  // Binop:
+  template <typename T0, typename T1, typename OpTy>
+  auto operator()(const T0 &Lhs, const T1 &Rhs, OpTy &&Op) {
+    static_assert(std::is_same_v<get_elem_type_t<T0>, get_elem_type_t<T1>>);
+    constexpr auto N = (std::max)(num_elements<T0>(), num_elements<T1>());
+    using ResultVec = vec<get_elem_type_t<T0>, N>;
+    return Op(static_cast<ResultVec>(Lhs), static_cast<ResultVec>(Rhs));
+  }
+  // Unary op:
+  template <typename T, typename OpTy> auto operator()(const T &X, OpTy &&Op) {
+    using ResultVec = vec<typename T::element_type, T::size()>;
+    return Op(static_cast<ResultVec>(X));
+  }
+};
+
+// Vector-specific part of the mixins' implementation.
+struct VectorImpl {
+private:
+#ifdef __SYCL_DEVICE_ONLY__
+  static constexpr bool is_host = false;
+#else
+  static constexpr bool is_host = true;
+#endif
+
+  template <typename T> static constexpr int num_elements() {
+    if constexpr (is_vec_or_swizzle_v<T>)
+      return T::size();
+    else
+      return 1;
+  }
+
+public:
+  // Binop:
+  template <typename T0, typename T1, typename OpTy>
+  auto operator()(const T0 &Lhs, const T1 &Rhs, OpTy &&Op) {
+    static_assert(std::is_same_v<get_elem_type_t<T0>, get_elem_type_t<T1>>);
+    constexpr auto N = (std::max)(num_elements<T0>(), num_elements<T1>());
+    using DataT = get_elem_type_t<T0>;
+    constexpr bool is_logical =
```

Comment:
mp_set_contains would be less verbose and is faster to compile (not sure whether it already matters for 8 elements in the set).